posit-dev/positron
# Add new version of "execute a command" that also waits

Addresses https://github.com/posit-dev/positron/issues/3256

You _can_ wait for commands in some situations to do something vs. only kick them off. An example is `'workbench.action.files.saveFiles'` which has an async accessor that waits through `editorService.saveAll()` and then returns success. This PR adds support for treating a command as an OpenRPC frontend request rather than an event. It basically adds back functionality that I took out in https://github.com/posit-dev/positron/pull/2425.

I think there are two options here:

- Have two versions of "execute a command", one which is a request (waits and returns something) and one which is an event (doesn't return). 
- Make "execute a command" always `await` because sometimes we must have that.

I'm a bit on the fence but tend to think I prefer having two versions? I definitely could be convinced either way.


### QA Notes

To see if we are really waiting:

- Clone the [ggplot2 repo](https://github.com/tidyverse/ggplot2) or other large R package and open it as a workspace
- Make a change in a file, like maybe adding something like `cat("Yeehaw!")` to one of the fortify methods like `fortify.lm` but DO NOT SAVE
- For extra credit, make a new untitled R file with some code in it, like `fortify(lm(mpg ~ ., data = mtcars), mtcars)` and DO NOT SAVE
- Use <kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> to load the dev version of the package
- The changed existing file should be saved, while the untitled file should _not_ be saved
- Run the code like `fortify(lm(mpg ~ ., data = mtcars), mtcars)`
- In the output, see "Yeehaw!"
